lib/workers.js

```diff
@@ -1,120 +1,97 @@
+ */
 
-We maintain two FIFO queues, one for available pythonCallers (if any)
-and one for waiting callbacks (if any). At any time at least one of
-these queues must be empty.
-
-Load is reported in:
-- python_worker_active: number of python workers currently processing jobs/callbacks
-- python_worker_idle: number of python workers available for incoming jobs/callbacks
-- python_callback_waiting: number of queued jobs/callbacks waiting for an available worker
-
-*/
-
-let pythonCallers = []; // all pythonCallers, whether available for work or currently busy
-let availablePythonCallers = []; // FIFO queue for available pythonCallers waiting for a job/callback
-let waitingCallbacks = []; // FIFO queue for jobs/callbacks waiting for an available pythonCaller
+/** @type {import('generic-pool').Pool<codeCaller.PythonCaller>} */
+let pool = null;
 
 module.exports = {
   init() {
     debug('init()');
-    if (!config.useWorkers) {
-      return;
-    }
-    var numWorkers = config.workersCount;
-    if (numWorkers == null) {
-      numWorkers = Math.ceil(config.workersPerCpu * os.cpus().length);
-    }
-    for (let i = 0; i < numWorkers; i++) {
-      const pc = new codeCaller.PythonCaller();
-      pc.number = i;
-      pythonCallers.push(pc);
-      load.startJob('python_worker_idle', pc.uuid);
-      availablePythonCallers.push(pc);
-    }
-    module.exports._warmUpWorkers();
+    const numWorkers = config.workersCount ?? Math.ceil(config.workersPerCpu * os.cpus().length);
+    pool = genericPool.createPool(
+      {
+        create: async () => {
+          const pc = new codeCaller.PythonCaller();
+          pc.ensureChild();
+          load.startJob('python_worker_idle', pc.uuid);
+          return pc;
+        },
+        destroy: async (pc) => {
+          pc.done();
+        },
+      },
+      {
+        min: numWorkers,
+        max: numWorkers,
+      }
+    );
+  },
+
+  async finishAsync() {
+    debug('finish(): waiting for pool to drain');
+    await pool.drain();
+    await pool.clear();
+    debug('finish(): pool finished draining');
   },
 
   finish(callback) {
-    debug('finish()');
-    if (!config.useWorkers) {
-      callback(null);
-      return;
-    }
-    const testFinished = () => {
-      if (availablePythonCallers.length < pythonCallers.length) {
-        // keep waiting for all pythonCallers to be returned
-        debug('finish(): waiting for availablePythonCallers');
-        setTimeout(testFinished, 100);
-      } else {
-        debug('finish(): calling done() on all pythonCallers');
-        for (let i = 0; i < pythonCallers.length; i++) {
-          pythonCallers[i].done();
-        }
-        pythonCallers = [];
-        availablePythonCallers = [];
-        return callback(null);
-      }
-    };
-    testFinished();
+    util.callbackify(module.exports.finishAsync)(callback);
   },
 
   getPythonCaller(callback) {
+    const jobUuid = uuidv4();
+    load.startJob('python_callback_waiting', jobUuid);
```

Comment:
There's a subtle change in behavior here. Before, if there was an available worker, we'd avoid recording load for `python_callback_waiting` entirely. Now, however, we'll momentarily report that load and then immediately end it. Based on the following line, that might end up reporting more load than desired:

https://github.com/PrairieLearn/PrairieLearn/blob/94bf9b5d8aaf8be782fbf92f7848efe97abe2e4f/lib/load.js#L73

One possible solution would be to switch to using [`hrtime`](https://nodejs.org/api/process.html#processhrtimebigint) to compute load deltas, which would free us from needing to round up from zero milliseconds. Or, we can just not worry about this?

Reply:
I'm think it should be fine to just not worry about it. Presumably 1 ms is significantly less time than it will take to process the request, so this should be just a small amount of noise on the signal.